### ISSUE TITLE
Do-it on a local variable gives nonsense value

### ISSUE THREAD

**Issue by halatmit** [mit-cml/appinventor-sources#281]
If you run do-it on a local varaible, you get a result like value:  "package $x" where x is the name of the variable.   If we can't make doiit work on locals, then at least we should give a sensible result.

**Comment by EnisM** [comment]
Or eliminate the DoIt for the local variables altogether.

**Comment by marksherman** [comment]
I know he's on vacation, but sounds like we need the opinion of @fturbak

**Comment by halatmit** [comment]
@jisqyv @mark-friedman 
Jeff:  I've been looking with this, and I'm mystified.   After playing around, it looks like if you evaluate an unbound variable foo in Kawa lexical environment, the value returned is the list (package foo). 
I'm wondering whether this somehow has to do with the companion.  Maybe we could look at it together.
Mark:  Any idea?

**Comment by jisqyv** [comment]
@halatmit You are correct. If you look at an unbound local variable in Kawa, you get the package object. I don't know if this behavior changes in a compiled environment.

-Jeff

**Comment by halatmit** [comment]
I've been playing more and this does not happen if you type directly at the
kawa-1.11 jar that is in our lib directory.
I wonder where the difference is coming from.



== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty


On Sat, Apr 18, 2015 at 10:06 PM, Jeffrey Schiller <notifications@github.com

**Comment by jisqyv** [comment]
I won't be able to play with this more until tomorrow. But I'm pretty sure
I was able to reproduce this with just the Kawa command line. I'll double
check tomorrow.

-Jeff

**Comment by halatmit** [comment]
Let me know how you did this with the Kawa command line.   I tried, and
Kawa worked fine -- but I hope you're right since that would make a lot
more sense.



== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty


On Sat, Apr 18, 2015 at 11:11 PM, Jeffrey Schiller <notifications@github.com
> wrote:

> I won't be able to play with this more until tomorrow. But I'm pretty sure
> I was able to reproduce this with just the Kawa command line. I'll double
> check tomorrow.
>
> -Jeff
>

**Comment by jisqyv** [comment]
@halatmit So I have not been able to reproduce this issue with just the Kawa command line. I have been looking over various source. When you do a "do it" on a lexical variable, the Yail "(lexical $var)" is sent to the companion. The Companion wraps this in a "delay" which is then wrapped with a Runnable which forces the resulting promise. This Runnable is then actually run in the Android UI thread via a call to the java "Handler:post". So it is hard to say what the environment looks like when the variable is finally evaluated...

**Comment by halatmit** [comment]
I'm playing with it.  Still investigating.



== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by halatmit** [comment]
tl;dr There's an issue with environments in the way we're using Kawa.

I've been looking at issue #281 -- trying do-it on a local variable FOO returns the (strange) result -- package FOO.   What you'd expect instead is an error, since FOO is unbound in the context of to do-it evaluation.

After messing around, it looks like any reference to an unbound variable behaves the same way:   referencing an unbound variable XXX returns package XXX. To demonstrate this, I generated an aia file, that has an unbound variable in the Yail, in the line:

(define-event Button1 Click()(set-this-form)
    (set-and-coerce-property! 'Label1 'Text HELLO!!!! 'text))

If you try this and press the button, you'll see the label text change to: package HELLO!!!!.

This does not happen if you build the app and run it compiled.   In that case, you get an error, but a strange one:
<unknown>internal compile error - caught
java.lang Unsupported Operation Except: can't load this type of class file.

This behavior also does not happen if you generate an unbound variable  at the Kawa command line -- even with the version of Kawa that is in our lib directory.  For that you get the expected unbound variable error.

Does anyone have an idea what might be going on.   (Mark?)

**Comment by halatmit** [comment]
EEK!
This seems to be an issue with Kawa and Android.   If you call Kawa's Scheme.eval with an unbound symbol like HELLO!!  from the App Inventor code running on the phone, the result is package HELLO!!!.
This does not happen when running Kawa from the command line.   So either we're not using the same Kawa, or there's something funny about Kawa in the Android environment.

Is the kawa jar in the App Inventor  lib directory, the same as what's compiled into App Inventor?

**Comment by jisqyv** [comment]
Yes, and I have been doing my tests using that jar file, so I am using the
same version.

At the moment I am in the depths of the Kawa source code, down the symbol
table hash tables...

-Jeff

On Monday, April 20, 2015, hal <notifications@github.com> wrote:

> EEK!
> This seems to be an issue with Kawa and Android. If you call Kawa's
> Scheme.eval with an unbound symbol like HELLO!! from the App Inventor code
> running on the phone, the result is package HELLO!!!.
> This does not happen when running Kawa from the command line. So either
> we're not using the same Kawa, or there's something funny about Kawa in the
> Android environment.
>
> Is the kawa jar in the App Inventor lib directory, the same as what's
> compiled into App Inventor?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-cml/appinventor-sources/issues/281#issuecomment-94572559>
> .
>


-- 
_______________________________________________________________________
Jeffrey I. Schiller
Information Systems and Technology – MIT App Inventor
Massachusetts Institute of Technology
77 Massachusetts Avenue  Room E17-110A, 32-392
Cambridge, MA 02139-4307
617.910.0259 - Voice
jis@mit.edu
http://jis.qyv.name
http://appinventor.mit.edu
_______________________________________________________________________

**Comment by halatmit** [comment]
You are a brave man to go into that code.


== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by halatmit** [comment]
I've been looking more and not making any progress.   What I've been trying to do is rebuild kawa from the source so I can have source code to play with.    I took kawa-1.11 and reapplied Mark's patches.  The system build fine.  Then I try to build app-inventor with the resulting kawa library a get a bunch of errors that think are happening in the pre-dexing stage.  That might just be another rathole to find out why.
I'm nervous that we apparently don't have the ability to rebuild the kawa library from source to get something that works with App Inventor.

**Comment by jisqyv** [comment]
What version of Java are you using?

-Jeff

On Tuesday, April 21, 2015, hal <notifications@github.com> wrote:

> I've been looking more and not making any progress. What I've been trying
> to do is rebuild kawa from the source so I can have source code to play
> with. I took kawa-1.11 and reapplied Mark's patches. The system build fine.
> Then I try to build app-inventor with the resulting kawa library a get a
> bunch of errors that think are happening in the pre-dexing stage. That
> might just be another rathole to find out why.
> I'm nervous that we apparently don't have the ability to rebuild the kawa
> library from source to get something that works with App Inventor.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-cml/appinventor-sources/issues/281#issuecomment-94881467>
> .
>


-- 
_______________________________________________________________________
Jeffrey I. Schiller
Information Systems and Technology – MIT App Inventor
Massachusetts Institute of Technology
77 Massachusetts Avenue  Room E17-110A, 32-392
Cambridge, MA 02139-4307
617.910.0259 - Voice
jis@mit.edu
http://jis.qyv.name
http://appinventor.mit.edu
_______________________________________________________________________

**Comment by halatmit** [comment]
I'm using java 1.7.0_45




== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by jisqyv** [comment]
I wonder if we have to "target" the compiler to generate java 1.6 bytecodes
in order for Dalvik to work.

-Jeff

**Comment by halatmit** [comment]
Let me know how to do that, and I'll play with it.

Right now, I'm clueless about what to try next.
And why dalvik would want 1.6 (if it does) is itself a mystery.



== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by marksherman** [comment]
Does it behave the same under ART?

**Comment by halatmit** [comment]
Liz says that her Qualcom code works with the latest version of dx.jar.  We should trey using that and see if Kawa now compiles and works in App Inventor.

**Comment by halatmit** [comment]
see PR #391

**Comment by jisqyv** [comment]
Fix/Workaround (#391) Merged into ucr
